syntax = "proto3";

package message;

option java_package = "io.pivotal.message";
option java_outer_classname = "MessageProtocol";

message Header {
    enum MessageType {
        UNSPECIFIED_TYPE = 0;
        CONNECT_REQUEST = 1;
        CONNECT_REPLY = 2;
        DISCONNECT_REQUEST = 3;
        DISCONNECT_REPLY = 4;
        PUT_REQUEST = 5;
        PUT_REPLY = 6;
        GET_REQUEST = 7;
        GET_REPLY = 8;
        INVALIDATE_REQUEST = 9;
        INVALIDATE_REPLY = 10;
        DESTROY_REQUEST = 11;
        DESTROY_REPLY = 12;
    }
    MessageType messageType = 1;
    fixed32 messageSize = 2;
    bool ok = 3;
}

message ConnectRequest {
    fixed32 id = 1;
}

message ConnectReply {
    fixed32 id = 1;
}

message DisconnectRequest {
    fixed32 id = 1;
}

message DisconnectReply {
    fixed32 id = 1;
}

message Pair {
    string key = 1;
    string value = 2;
}

message PutRequest {
    string region = 1;
    repeated Pair pair = 2;
}

message PutReply {
    fixed32 count = 1;
}

message GetRequest {
    string region = 1;
    repeated string key = 2;
}

message GetReply {
    repeated Pair pair = 1;
}

message InvalidateRequest {
    string region = 1;
    repeated string key = 2;
}

message InvalidateReply {
    fixed32 count = 1;
}

message DestroyRequest {
    string region = 1;
    repeated string key = 2;
}

message DestroyReply {
    fixed32 count = 1;
}
